### DISCUSSION TITLE
How to make header value available in subscription function?

### DISCUSSION THREAD

**Discussion by pubrrr** [async-graphql/async-graphql#1113] [Q&A] [answered]
Hi,

for queries and mutations it's possible to pass data to resolver functions via `schema.execute(request.data(my_data))`. Is something similar possible for subscriptions? I'm using the warp integration. Probably I could do it by reimplementing `async_graphql_warp::graphql_subscription`, but by using warp so far, I feel like there should be something that works more out of the box?

The specific use case is: I want to have a header value / cookie available in my subscription resolver function.

**Comment by sunli829** [comment, marked as answer]
Use [GraphQLWebSocket](https://docs.rs/async-graphql-warp/4.0.16/async_graphql_warp/struct.GraphQLWebSocket.html) instead of `async_graphql_warp::graphql_subscription`, which provides more customization options.

```rust
let filter = warp::ws()
    .and(graphql_protocol())
    .map(move |ws: ws::Ws, protocol| {
        let schema = schema.clone();

        let reply = ws
            .on_upgrade(move |socket| {
               let ws = GraphQLWebSocket::new(socket, schema, protocol).with_data(...); // <<<<<<
               ws.serve()
            });

        warp::reply::with_header(
            reply,
            "Sec-WebSocket-Protocol",
            protocol.sec_websocket_protocol(),
        )
    });
```
